[JSC][32bit] Enable concurrent compilation

mikhailramalho opened the pull request:
#### df50b57bbbf11f79296ed7298983d930bf787496
<pre>
[JSC][32bit] Enable concurrent compilation
<a href="https://bugs.webkit.org/show_bug.cgi?id=239821">https://bugs.webkit.org/show_bug.cgi?id=239821</a>

This enables concurrent compilation in 32-bit systems. The patch increases
JetStream2's score by 9.2%, compared to the sequential compilation version
(ToT on ARMv7).

The patch works by checking if a cell is live whenever we try to get a
cell's type from the payload, and adds a new watchpoint to DFG, that
jettisons the compiled code if a cell used during compilation is found to be dead.

The check if a cell is live is divided into two steps:
(1) We check if the cell pointer is valid, i.e., if it is part of the
precise allocation set, or if its markedBlock is part of the marked block set.
This required changes to make both sets thread-safe. There was a small regression
in performance in JetStream's score in 32bit systems using sequential compilation
(-0.4%) but no regression in intel 64bit linux.
(2) We check if the cell is live, similar to HeapCell:isLive. Special case for
when a cell's markedBlock is freelisted: the cell would be in a dead-but-not-destructed
state, so we check if it was zapped.

This patch also includes:
(1) Rewrite of MarkedBlockSet to be thread-safe and the creation of thread-safe
PreciseAllocationSet. This required a rewrite of some functions in HeapUtil and
its users.
(2) Creation of the singleton CellAddressChecker: it holds a set of heaps, used
to check if a pointer is actually a cell. I tried to use thread_local storage but
it caused JSC to run OOM in some tests.
(3) Some changes to DFG's abstract interpreter so it doesn't crash when it finds
a dead cell: a dead cell sets the abstract value to clear(), which
AbstractValue::checkConsistency() check that if the abstract value is a cell, it
should never be clear. The changes here check the cell type and clear the abstract
value, and don't call AbstractValue::checkConsistency() afterward.

JetStream2 results:

-----------------------------------------------------------------------------------------------------------------------
|          subtest          |    pts      |    pts      |  b / a   | pValue (significance using False Discovery Rate) |
-----------------------------------------------------------------------------------------------------------------------
| 3d-cube-SP                |166.842257   |191.651250   |1.148697  | 0.000000 (significant)                           |
| 3d-raytrace-SP            |121.830845   |137.467771   |1.128349  | 0.000000 (significant)                           |
| Air                       |94.564043    |113.849099   |1.203936  | 0.000000 (significant)                           |
| Babylon                   |136.550170   |197.096390   |1.443399  | 0.000000 (significant)                           |
| Basic                     |145.174218   |172.201772   |1.186173  | 0.000000 (significant)                           |
| Box2D                     |96.168053    |135.336668   |1.407293  | 0.000000 (significant)                           |
| FlightPlanner             |179.733139   |218.052388   |1.213201  | 0.000000 (significant)                           |
| ML                        |22.580790    |23.779379    |1.053080  | 0.000000 (significant)                           |
| UniPoker                  |77.069218    |76.944716    |0.998385  | 0.745485                                         |
| ai-astar                  |172.339827   |182.194075   |1.057179  | 0.000000 (significant)                           |
| async-fs                  |58.373134    |66.001618    |1.130685  | 0.000000 (significant)                           |
| base64-SP                 |163.675701   |164.258499   |1.003561  | 0.386329                                         |
| cdjs                      |64.786358    |68.880058    |1.063188  | 0.000000 (significant)                           |
| chai-wtb                  |30.723964    |35.512906    |1.155870  | 0.000000 (significant)                           |
| crypto                    |370.774651   |422.601554   |1.139780  | 0.000000 (significant)                           |
| crypto-aes-SP             |166.297273   |180.887002   |1.087733  | 0.000000 (significant)                           |
| crypto-md5-SP             |138.726334   |144.764783   |1.043528  | 0.000000 (significant)                           |
| crypto-sha1-SP            |148.701098   |154.401856   |1.038337  | 0.000000 (significant)                           |
| date-format-tofte-SP      |93.721970    |95.972233    |1.024010  | 0.000006 (significant)                           |
| date-format-xparb-SP      |84.335641    |84.399444    |1.000757  | 0.884345                                         |
| delta-blue                |203.552572   |250.170209   |1.229020  | 0.000000 (significant)                           |
| earley-boyer              |183.528931   |267.870222   |1.459553  | 0.000000 (significant)                           |
| first-inspector-code-load |130.165322   |125.707273   |0.965751  | 0.000000 (significant)                           |
| gaussian-blur             |105.136392   |102.088203   |0.971007  | 0.000000 (significant)                           |
| gbemu                     |29.254298    |34.314988    |1.172990  | 0.000000 (significant)                           |
| hash-map                  |128.335637   |125.664184   |0.979184  | 0.000003 (significant)                           |
| jshint-wtb                |13.149089    |14.809269    |1.126258  | 0.000000 (significant)                           |
| json-parse-inspector      |83.706580    |92.194498    |1.101401  | 0.000000 (significant)                           |
| json-stringify-inspector  |108.521622   |104.554735   |0.963446  | 0.000008 (significant)                           |
| multi-inspector-code-load |139.224604   |136.393483   |0.979665  | 0.000000 (significant)                           |
| n-body-SP                 |336.090961   |340.670871   |1.013627  | 0.000025 (significant)                           |
| navier-stokes             |425.875692   |406.026662   |0.953392  | 0.000000 (significant)                           |
| octane-code-load          |485.779704   |484.639128   |0.997652  | 0.708012                                         |
| pdfjs                     |51.953615    |59.042339    |1.136443  | 0.000000 (significant)                           |
| prepack-wtb               |12.559819    |16.765718    |1.334869  | 0.000000 (significant)                           |
| raytrace                  |184.737268   |214.173299   |1.159340  | 0.000000 (significant)                           |
| regex-dna-SP              |217.026089   |210.995824   |0.972214  | 0.000000 (significant)                           |
| regexp                    |120.429176   |122.785703   |1.019568  | 0.000459 (significant)                           |
| richards                  |231.044982   |248.550770   |1.075768  | 0.000000 (significant)                           |
| splay                     |94.720294    |115.778294   |1.222318  | 0.000000 (significant)                           |
| stanford-crypto-aes       |108.887712   |108.180925   |0.993509  | 0.205676                                         |
| stanford-crypto-pbkdf2    |210.891098   |225.484971   |1.069201  | 0.000000 (significant)                           |
| stanford-crypto-sha256    |190.608578   |190.512030   |0.999493  | 0.885449                                         |
| string-unpack-code-SP     |144.199974   |145.742765   |1.010699  | 0.001692 (significant)                           |
| tagcloud-SP               |117.106005   |119.112579   |1.017135  | 0.000080 (significant)                           |
-----------------------------------------------------------------------------------------------------------------------

a mean = 114.27907
b mean = 124.75480
pValue = 0.0000000000
(Bigger means are better.)
1.092 times better
Results ARE significant

* JSTests/stress/dont-link-virtual-calls-on-compiler-thread.js:
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/bytecode/AccessCase.cpp:
(JSC::AccessCase::generateImpl):
* Source/JavaScriptCore/bytecode/SpeculatedType.cpp:
(JSC::speculationFromCell):
* Source/JavaScriptCore/bytecode/Watchpoint.h:
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGAbstractValue.cpp:
(JSC::DFG::AbstractValue::set):
(JSC::DFG::AbstractValue::mergeOSREntryValue):
* Source/JavaScriptCore/dfg/DFGAbstractValue.h:
(JSC::DFG::AbstractValue::validateOSREntryValue const):
* Source/JavaScriptCore/dfg/DFGDesiredWatchpoints.cpp:
(JSC::DFG::JSCellWatchpointAdaptor::add):
(JSC::DFG::DesiredWatchpoints::addLazily):
(JSC::DFG::DesiredWatchpoints::reallyAdd):
(JSC::DFG::DesiredWatchpoints::areStillValid const):
(JSC::DFG::DesiredWatchpoints::dumpInContext const):
* Source/JavaScriptCore/dfg/DFGDesiredWatchpoints.h:
(JSC::DFG::JSCellWatchpointAdaptor::hasBeenInvalidated):
(JSC::DFG::JSCellWatchpointAdaptor::dumpInContext):
(JSC::DFG::DesiredWatchpoints::dump const):
* Source/JavaScriptCore/dfg/DFGDesiredWeakReferences.h:
* Source/JavaScriptCore/dfg/DFGGraph.cpp:
(JSC::DFG::Graph::freeze):
* Source/JavaScriptCore/dfg/DFGGraph.h:
* Source/JavaScriptCore/dfg/DFGPlan.cpp:
(JSC::DFG::Plan::finalize):
* Source/JavaScriptCore/heap/ConservativeRoots.cpp:
(JSC::ConservativeRoots::genericAddPointer):
(JSC::ConservativeRoots::genericAddSpan):
* Source/JavaScriptCore/heap/ConservativeRoots.h:
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::Heap):
(JSC::Heap::~Heap):
(JSC::CellAddressChecker::instance):
(JSC::CellAddressChecker::add):
(JSC::CellAddressChecker::remove):
(JSC::CellAddressChecker::isValidCell):
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/heap/HeapUtil.h:
(JSC::HeapUtil::findGCObjectPointersForMarking):
(JSC::HeapUtil::isPointerGCObjectJSCell):
(JSC::HeapUtil::isValueGCObject):
* Source/JavaScriptCore/heap/MarkedBlockSet.h:
(JSC::MarkedBlockSet::add):
(JSC::MarkedBlockSet::remove):
(JSC::MarkedBlockSet::size):
(JSC::MarkedBlockSet::contains):
(JSC::MarkedBlockSet::forEachLiveCell):
(JSC::MarkedBlockSet::forEachDeadCell):
(JSC::MarkedBlockSet::filterRuleOut):
(JSC::MarkedBlockSet::filter const): Deleted.
(JSC:: const): Deleted.
* Source/JavaScriptCore/heap/MarkedSpace.cpp:
(JSC::MarkedSpace::~MarkedSpace):
(JSC::MarkedSpace::enablePreciseAllocationTracking):
* Source/JavaScriptCore/heap/MarkedSpace.h:
(JSC::MarkedSpace::preciseAllocationSet const):
* Source/JavaScriptCore/heap/MarkedSpaceInlines.h:
(JSC::MarkedSpace::forEachLiveCell):
(JSC::MarkedSpace::forEachDeadCell):
* Source/JavaScriptCore/jit/AssemblyHelpers.h:
(JSC::AssemblyHelpers::storeCell):
* Source/JavaScriptCore/jit/JITPlan.cpp:
(JSC::verboseCompilationEnabled):
(JSC::JITPlan::compileInThread):
* Source/JavaScriptCore/runtime/JSCell.cpp:
(JSC::isLiveConcurrently):
* Source/JavaScriptCore/runtime/JSCell.h:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/runtime/SamplingProfiler.cpp:
(JSC::SamplingProfiler::processUnverifiedStackTraces):
* Source/JavaScriptCore/tools/VMInspector.cpp:
(JSC::VMInspector::isInHeap):
* Source/WTF/wtf/PlatformEnable.h:
* </pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/356d1079ad3e8cba5879cc9158dfcaf0f716f7a7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105846 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14914 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38694 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115044 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/175173 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109748 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16333 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6215 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98079 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114817 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111595 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12435 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95403 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39977 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94295 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27045 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81636 "Found 1 new API test failure: /WebKitGTK/TestWebViewEditor:/webkit/WebKitWebView/cut-copy-paste/editable (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/95539 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8185 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28397 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/94824 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/5982 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8667 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5058 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30448 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14294 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47945 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/103571 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10224 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25680 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/JSCell.cpp:
```diff
@@ -338,4 +338,33 @@ NEVER_INLINE NO_RETURN_DUE_TO_CRASH NOT_TAIL_CALLED void reportZappedCellAndCras
+    if (!candidate)
+        return false;
+
+    if (cell->isPreciseAllocation())
+        return cell->preciseAllocation().isLive();
```
This is not correct. If random precise-allocation-like value comes, then this will crash.

mikhailramalho replied:
I didn't get it, could you elaborate? This is the same code as in `HeapCell::isLive()`.

Constellation replied:
> I didn't get it, could you elaborate? This is the same code as in HeapCell::isLive().

No. HeapCell::isLive is OK because it is called after ensuring that this is a valid HeapCell address. This function is not guaranteeing that.

mikhailramalho replied:
Alright, does it mean that I need to do something like `isPointerGCObjectJSCell`?
```
        if (pointer->isPreciseAllocation()) {
            auto* set = heap.objectSpace().preciseAllocationSet();
            ASSERT(set);
            if (set->isEmpty())
                return false;
            if (!set->isValidValue(pointer))
                return false;
            return set->contains(pointer);
        }
```
I tried that but the `set` I get is always `0x0`.

mikhailramalho replied:
I added guards to try to retrieve the structure IDs, WDYT?

mikhailramalho replied:
Got that from `HeapVerifier::validateJSCell`

mikhailramalho replied:
I added  `ASSERT(!cell->isZapped());` and `ASSERT(cell->structureID());`, is it enough to make sure the pointer is a cell, right?

MenloDorian replied:
You cannot just lift code from `HeapVerifier`.  `HeapVerifier` runs while the world is stopped, and it does not have to deal with concurrency issues.

To implement this feature, you need to understand the GC architecture very well at a fundamental level and think of the concurrency issues.

Constellation commented on Source/JavaScriptCore/runtime/JSCell.cpp:
```diff
@@ -338,4 +338,33 @@ NEVER_INLINE NO_RETURN_DUE_TO_CRASH NOT_TAIL_CALLED void reportZappedCellAndCras
+    if (cell->isPreciseAllocation())
+        return cell->preciseAllocation().isLive();
+
+    Locker locker { candidate->lock() };
```
This is not correct since there is no guarantee that candidate is a valid MarkedBlock pointer.

mikhailramalho replied:
We use `MarkedBlock::blockFor(cell)` to get the right candidate, wouldn't it prevent us from getting an invalid MarkedBlock pointer?

Constellation replied:
It does not prevent us from getting an invalid MarkedBlock pointer.

mikhailramalho replied:
Here I added the filter to ruleOut bad candidates, like it's done in `isPointerGCObjectJSCell`

Constellation commented on Source/JavaScriptCore/bytecode/AccessCase.cpp:
```diff
@@ -2100,6 +2100,7 @@ void AccessCase::generateImpl(AccessGenerationState& state)
                 CCallHelpers::TrustedImm32(numberOfParameters),
                 calleeFrame.withOffset(CallFrameSlot::argumentCountIncludingThis * sizeof(Register) + PayloadOffset));
 
+            // TODO: in 32bit platforms we spend an extra instruction to move the CellTag on the second storeCell unnecessarily
```
We do not use `TODO`. Use `FIXME`.

mikhailramalho replied:
done.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerMIPS.h:
```diff
@@ -1358,6 +1357,12 @@ class MacroAssemblerMIPS : public AbstractMacroAssembler<Assembler> {
         }
     }
 
+    void loadPair32(const void* address, RegisterID dest1, RegisterID dest2)
```
Do not add `const void*` one newly. Use `AbsoluteAddress`.

mikhailramalho replied:
done.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h:
```diff
@@ -982,6 +982,11 @@ class MacroAssemblerARMv7 : public AbstractMacroAssembler<Assembler> {
         loadPair32(Address(scratch, address.offset), dest1, dest2);
     }
 
+    void loadPair32(const void* address, RegisterID dest1, RegisterID dest2)
```
Ditto.

mikhailramalho replied:
done.

Constellation commented on Source/JavaScriptCore/heap/ConservativeRoots.cpp:
```diff
@@ -94,11 +94,10 @@ void ConservativeRoots::genericAddSpan(void* begin, void* end, MarkHook& markHoo
     HeapVersion newlyAllocatedVersion = m_heap.objectSpace().newlyAllocatedVersion();
     for (char** it = static_cast<char**>(begin); it != static_cast<char**>(end); ++it)
-        genericAddPointer(*it, markingVersion, newlyAllocatedVersion, filter, markHook);
+        genericAddPointer(*it, markingVersion, newlyAllocatedVersion, markHook);
```
Why is this change necessary?

mikhailramalho replied:
No, but I thought it was a worth addition. I'll revert it.

Constellation commented on Source/JavaScriptCore/heap/HeapUtil.h:
```diff
@@ -46,8 +46,7 @@ class HeapUtil {
     static void findGCObjectPointersForMarking(
-        Heap& heap, HeapVersion markingVersion, HeapVersion newlyAllocatedVersion, TinyBloomFilter<uintptr_t> filter,
-        void* passedPointer, const Func& func)
+        Heap& heap, HeapVersion markingVersion, HeapVersion newlyAllocatedVersion, void* passedPointer, const Func& func)
```
Ditto, why is this change necessary?

Constellation commented on Source/JavaScriptCore/heap/HeapUtil.h:
```diff
@@ -81,6 +80,8 @@ class HeapUtil {
     
         MarkedBlock* candidate = MarkedBlock::blockFor(pointer);
+        TinyBloomFilter<uintptr_t> filter = heap.objectSpace().blocks().filter();
+
```
Ditto, why is this change necessary?

Constellation commented on Source/JavaScriptCore/heap/HeapUtil.h:
```diff
@@ -134,7 +135,7 @@ class HeapUtil {
     }
     
-    static bool isPointerGCObjectJSCell(Heap& heap, TinyBloomFilter<uintptr_t> filter, JSCell* pointer)
+    static bool isPointerGCObjectJSCell(Heap& heap, JSCell* pointer)
```
Ditto, why is this change necessary?

Constellation commented on Source/JavaScriptCore/heap/HeapUtil.h:
```diff
@@ -152,8 +153,9 @@ class HeapUtil {
-        
         MarkedBlock* candidate = MarkedBlock::blockFor(pointer);
+
+        TinyBloomFilter<uintptr_t> filter = heap.objectSpace().blocks().filter();
```
Ditto, why is this change necessary?

Constellation commented on Source/JavaScriptCore/heap/HeapUtil.h:
```diff
@@ -175,13 +177,12 @@ class HeapUtil {
     // This does not find the cell if the pointer is pointing at the middle of a JSCell.
-    static bool isValueGCObject(
-        Heap& heap, TinyBloomFilter<uintptr_t> filter, JSValue value)
+    static bool isValueGCObject(Heap& heap, JSValue value)
```
Ditto, why is this change necessary?

Constellation commented on Source/JavaScriptCore/heap/HeapUtil.h:
```diff
@@ -175,13 +177,12 @@ class HeapUtil {
         if (!value.isCell())
             return false;
-        return isPointerGCObjectJSCell(heap, filter, value.asCell());
+        return isPointerGCObjectJSCell(heap, value.asCell());
```
Ditto, why is this change necessary?

Constellation commented on Source/JavaScriptCore/heap/MarkedBlock.cpp:
```diff
@@ -107,6 +107,7 @@ MarkedBlock::Header::~Header()
 
 void MarkedBlock::Handle::unsweepWithNoNewlyAllocated()
 {
+    Locker locker { blockHeader().m_lock };
```
Why is this change necessary?

mikhailramalho committed:
[JSC][32bit] Enable concurrency
https://bugs.webkit.org/show_bug.cgi?id=239821

Reviewed by NOBODY (OOPS!).

This allows concurrent compilation in 32-bit systems (disabled by default for now).

The patch works by checking if a cell is live whenever we try to get a
cell's type from the payload, and adds a new watchpoint to DFG, that
jettisons the compiled code if a cell is found to be dead.

To check if a cell is live is divided into two steps:
(1) We check if the cell pointer is valid, i.e., if it is part of the
precise allocation set, or if its markedBlock is part of the marked block set.
This required changes to make both sets thread-safe. There was a small regression
in performance in JetStream's score in 32bit systems using sequential compilation
(-0.4%) but no regression in intel 64bit linux.
(2) We check if the cell is live, similar to HeapCell:isLive. Special case for
when a cell's markedBlock is freelisted: the cell would be in a dead-but-not-destructed
state, so we check if it was zapped.

This patch also includes:
(1) Rewrite of MarkedBlockSet to be thread-safe and the creation of thread-safe
PreciseAllocationSet. This required a rewrite of some functions in HeapUtil and
its users.
(2) Creation of the singleton CellAddressChecker: it holds a set of heaps, used
to check if a pointer is actually a cell. I tried to use thread_local storage but
it caused JSC to run OOM in some tests.
(3) Some changes to DFG's abstract interpreter so it doesn't crash when it finds
a dead cell.

The patch increases JetStream2's core by 9.2%, compared to the sequential
compilation version (ToT on ARMv7):

-----------------------------------------------------------------------------------------------------------------------
|          subtest          |    pts      |    pts      |  b / a   | pValue (significance using False Discovery Rate) |
-----------------------------------------------------------------------------------------------------------------------
| 3d-cube-SP                |166.842257   |191.651250   |1.148697  | 0.000000 (significant)                           |
| 3d-raytrace-SP            |121.830845   |137.467771   |1.128349  | 0.000000 (significant)                           |
| Air                       |94.564043    |113.849099   |1.203936  | 0.000000 (significant)                           |
| Babylon                   |136.550170   |197.096390   |1.443399  | 0.000000 (significant)                           |
| Basic                     |145.174218   |172.201772   |1.186173  | 0.000000 (significant)                           |
| Box2D                     |96.168053    |135.336668   |1.407293  | 0.000000 (significant)                           |
| FlightPlanner             |179.733139   |218.052388   |1.213201  | 0.000000 (significant)                           |
| ML                        |22.580790    |23.779379    |1.053080  | 0.000000 (significant)                           |
| UniPoker                  |77.069218    |76.944716    |0.998385  | 0.745485                                         |
| ai-astar                  |172.339827   |182.194075   |1.057179  | 0.000000 (significant)                           |
| async-fs                  |58.373134    |66.001618    |1.130685  | 0.000000 (significant)                           |
| base64-SP                 |163.675701   |164.258499   |1.003561  | 0.386329                                         |
| cdjs                      |64.786358    |68.880058    |1.063188  | 0.000000 (significant)                           |
| chai-wtb                  |30.723964    |35.512906    |1.155870  | 0.000000 (significant)                           |
| crypto                    |370.774651   |422.601554   |1.139780  | 0.000000 (significant)                           |
| crypto-aes-SP             |166.297273   |180.887002   |1.087733  | 0.000000 (significant)                           |
| crypto-md5-SP             |138.726334   |144.764783   |1.043528  | 0.000000 (significant)                           |
| crypto-sha1-SP            |148.701098   |154.401856   |1.038337  | 0.000000 (significant)                           |
| date-format-tofte-SP      |93.721970    |95.972233    |1.024010  | 0.000006 (significant)                           |
| date-format-xparb-SP      |84.335641    |84.399444    |1.000757  | 0.884345                                         |
| delta-blue                |203.552572   |250.170209   |1.229020  | 0.000000 (significant)                           |
| earley-boyer              |183.528931   |267.870222   |1.459553  | 0.000000 (significant)                           |
| first-inspector-code-load |130.165322   |125.707273   |0.965751  | 0.000000 (significant)                           |
| gaussian-blur             |105.136392   |102.088203   |0.971007  | 0.000000 (significant)                           |
| gbemu                     |29.254298    |34.314988    |1.172990  | 0.000000 (significant)                           |
| hash-map                  |128.335637   |125.664184   |0.979184  | 0.000003 (significant)                           |
| jshint-wtb                |13.149089    |14.809269    |1.126258  | 0.000000 (significant)                           |
| json-parse-inspector      |83.706580    |92.194498    |1.101401  | 0.000000 (significant)                           |
| json-stringify-inspector  |108.521622   |104.554735   |0.963446  | 0.000008 (significant)                           |
| multi-inspector-code-load |139.224604   |136.393483   |0.979665  | 0.000000 (significant)                           |
| n-body-SP                 |336.090961   |340.670871   |1.013627  | 0.000025 (significant)                           |
| navier-stokes             |425.875692   |406.026662   |0.953392  | 0.000000 (significant)                           |
| octane-code-load          |485.779704   |484.639128   |0.997652  | 0.708012                                         |
| pdfjs                     |51.953615    |59.042339    |1.136443  | 0.000000 (significant)                           |
| prepack-wtb               |12.559819    |16.765718    |1.334869  | 0.000000 (significant)                           |
| raytrace                  |184.737268   |214.173299   |1.159340  | 0.000000 (significant)                           |
| regex-dna-SP              |217.026089   |210.995824   |0.972214  | 0.000000 (significant)                           |
| regexp                    |120.429176   |122.785703   |1.019568  | 0.000459 (significant)                           |
| richards                  |231.044982   |248.550770   |1.075768  | 0.000000 (significant)                           |
| splay                     |94.720294    |115.778294   |1.222318  | 0.000000 (significant)                           |
| stanford-crypto-aes       |108.887712   |108.180925   |0.993509  | 0.205676                                         |
| stanford-crypto-pbkdf2    |210.891098   |225.484971   |1.069201  | 0.000000 (significant)                           |
| stanford-crypto-sha256    |190.608578   |190.512030   |0.999493  | 0.885449                                         |
| string-unpack-code-SP     |144.199974   |145.742765   |1.010699  | 0.001692 (significant)                           |
| tagcloud-SP               |117.106005   |119.112579   |1.017135  | 0.000080 (significant)                           |
-----------------------------------------------------------------------------------------------------------------------

a mean = 114.27907
b mean = 124.75480
pValue = 0.0000000000
(Bigger means are better.)
1.092 times better
Results ARE significant

* JSTests/stress/dont-link-virtual-calls-on-compiler-thread.js:
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/bytecode/AccessCase.cpp:
(JSC::AccessCase::generateImpl):
* Source/JavaScriptCore/bytecode/SpeculatedType.cpp:
(JSC::speculationFromCell):
* Source/JavaScriptCore/bytecode/Watchpoint.h:
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGAbstractValue.cpp:
(JSC::DFG::AbstractValue::set):
(JSC::DFG::AbstractValue::mergeOSREntryValue):
* Source/JavaScriptCore/dfg/DFGAbstractValue.h:
(JSC::DFG::AbstractValue::validateOSREntryValue const):
* Source/JavaScriptCore/dfg/DFGDesiredWatchpoints.cpp:
(JSC::DFG::JSCellWatchpointAdaptor::add):
(JSC::DFG::DesiredWatchpoints::addLazily):
(JSC::DFG::DesiredWatchpoints::reallyAdd):
(JSC::DFG::DesiredWatchpoints::areStillValid const):
(JSC::DFG::DesiredWatchpoints::dumpInContext const):
* Source/JavaScriptCore/dfg/DFGDesiredWatchpoints.h:
(JSC::DFG::JSCellWatchpointAdaptor::hasBeenInvalidated):
(JSC::DFG::JSCellWatchpointAdaptor::dumpInContext):
* Source/JavaScriptCore/dfg/DFGGraph.cpp:
(JSC::DFG::Graph::freeze):
* Source/JavaScriptCore/dfg/DFGGraph.h:
* Source/JavaScriptCore/dfg/DFGPlan.cpp:
(JSC::DFG::Plan::finalize):
* Source/JavaScriptCore/heap/ConservativeRoots.cpp:
(JSC::ConservativeRoots::genericAddPointer):
(JSC::ConservativeRoots::genericAddSpan):
* Source/JavaScriptCore/heap/ConservativeRoots.h:
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::Heap):
(JSC::Heap::~Heap):
(JSC::CellAddressChecker::instance):
(JSC::CellAddressChecker::add):
(JSC::CellAddressChecker::remove):
(JSC::CellAddressChecker::isValidCell):
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/heap/HeapUtil.h:
(JSC::HeapUtil::findGCObjectPointersForMarking):
(JSC::HeapUtil::isPointerGCObjectJSCell):
(JSC::HeapUtil::isValueGCObject):
* Source/JavaScriptCore/heap/MarkedBlockSet.h:
(JSC::MarkedBlockSet::add):
(JSC::MarkedBlockSet::remove):
(JSC::MarkedBlockSet::size):
(JSC::MarkedBlockSet::contains):
(JSC::MarkedBlockSet::forEachLiveCell):
(JSC::MarkedBlockSet::forEachDeadCell):
(JSC::MarkedBlockSet::filterRuleOut):
(JSC::MarkedBlockSet::filter const): Deleted.
(JSC:: const): Deleted.
* Source/JavaScriptCore/heap/MarkedSpace.cpp:
(JSC::MarkedSpace::~MarkedSpace):
(JSC::MarkedSpace::enablePreciseAllocationTracking):
* Source/JavaScriptCore/heap/MarkedSpace.h:
(JSC::MarkedSpace::preciseAllocationSet const):
* Source/JavaScriptCore/heap/MarkedSpaceInlines.h:
(JSC::MarkedSpace::forEachLiveCell):
(JSC::MarkedSpace::forEachDeadCell):
* Source/JavaScriptCore/jit/AssemblyHelpers.h:
(JSC::AssemblyHelpers::storeCell):
* Source/JavaScriptCore/jit/JITPlan.cpp:
(JSC::verboseCompilationEnabled):
(JSC::JITPlan::compileInThread):
* Source/JavaScriptCore/runtime/JSCell.cpp:
(JSC::isLiveConcurrently):
* Source/JavaScriptCore/runtime/JSCell.h:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/runtime/SamplingProfiler.cpp:
(JSC::SamplingProfiler::processUnverifiedStackTraces):
* Source/JavaScriptCore/tools/VMInspector.cpp:
(JSC::VMInspector::isInHeap):
* Source/WTF/wtf/PlatformEnable.h:

jjgriego commented on Source/JavaScriptCore/dfg/DFGAbstractValue.cpp:
```diff
@@ -50,30 +50,32 @@ void AbstractValue::observeTransitions(const TransitionVector& vector)
-    
-    checkConsistency();
+        m_value = value.value();
+        checkConsistency();
```
why not `checkConsistency` before returning in all branches, like before?

mikhailramalho replied:
`clear()` already calls `checkConsistency()`.

jjgriego commented on Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
```diff
@@ -440,7 +440,8 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
         // Assert that the state of arguments has been set. SetArgumentDefinitely/SetArgumentMaybe means
         // that someone set the argument values out-of-band, and currently this always means setting to a
         // non-clear value.
```
(this comment is out of date if we remove the assertion)
